src/main/java/com/mhkim/tms/auth/oauth/COAuth2UserService.java

```diff
@@ -0,0 +1,60 @@
+package com.mhkim.tms.auth.oauth;
+
+import java.util.Collections;
+
+import javax.servlet.http.HttpSession;
+
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserService;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+
+import com.mhkim.tms.v1.user.dto.SessionUser;
+import com.mhkim.tms.v1.user.entity.User;
+import com.mhkim.tms.v1.user.repository.UserRepository;
+
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Custom COAuth2UserService
+ */
+@RequiredArgsConstructor
+@Service
+public class COAuth2UserService implements OAuth2UserService<OAuth2UserRequest, OAuth2User> {
+    
+    private final UserRepository userRepository;
+    private final HttpSession httpSession;
+
+    @Override
+    public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+        OAuth2User oAuth2User = new DefaultOAuth2UserService().loadUser(userRequest);
+
+        String registrationId = userRequest.getClientRegistration().getRegistrationId();
+        String userNameAttributeName = userRequest.getClientRegistration().getProviderDetails().getUserInfoEndpoint().getUserNameAttributeName();
+        
+        OAuthAttributes attributes = OAuthAttributes.of(oAuth2User.getAttributes(), registrationId, userNameAttributeName);
+        
+        User user = updateUser(attributes);
+        httpSession.setAttribute("user", new SessionUser(user));
+
+        return new DefaultOAuth2User(
+                Collections.singleton(new SimpleGrantedAuthority(user.getRoleKey())),
+                attributes.getAttributes(),
+                attributes.getNameAttributeKey());
+    }
+
+    private User updateUser(OAuthAttributes attributes) {
+        User user = userRepository.findByEmail(attributes.getEmail())
+                .map(u -> {
+                    u.updateUser(attributes.getName());
+                    return u;
+                }).orElse(attributes.toEntity());
+
+        return userRepository.save(user);
```

Comment:
```suggestion
        return userRepository.findByEmail(attributes.getEmail())
                .map(user -> {
                    user.updateUser(attributes.getName());
                    return userRepository.save(user);
                }).orElse(attributes.toEntity());

```
여기서 return userRepository.save(user); 도 map 안으로 넣어주면 더 깔끔해질 것 같아요.

Reply:
메소드명 네이밍 떄문에 오해할 수 있을듯
insert 와 update 를 동시에 하는 기능이라 소스는 기존대로 유지하고 
메소드명을 saveUser()로  변경!